[PROF-4779] Refactor in prepraration for use of libddprof from profiling native extension

ivoanjo opened the pull request:
[`libddprof`](https://github.com/DataDog/libddprof) is a Rust-based library that implements common functionality used by our profilers:

* aggregation of profiling data
* serialization of profiling data
* reporting of profiling data to backend

This PR adds the scaffolding to enable usage of libddprof from dd-trace-rb, **but does not yet make use of any of its functionality**.

Currently, libddprof works for all supported Rubies (2.1 to 3.2 dev) but only on x86-64 and 64-bit ARM Linux. The current integration has been set up in such a way that on other platforms we print a nice banner telling customers that they won't have profiling **but do not otherwise impact the customer in using all other Datadog products** (tracing, ci, appsec, ...).

This PR includes a nice refactoring + addition of tests to make sure this safe installation behavior is working correctly.

**Note**: To allow this PR to be merged in more quickly, I've commented out the actual libddprof linking bits in `extconf.rb`; I'll re-enable that in a later PR.

lloeki reviewed (approved):
LGTM, probably worth another pair of eyes as to how this behaves overall WRT general ddtrace user usage.

ivoanjo commented:
This PR was opened to target the 1.0 branch. I've just rebased it now on top of current master, and moved it to target that branch, since 1.0 is no more. No further changes :)

ivoanjo commented:
I've just did another rebase of this PR, fixing a conflict and adding a few cleanups for things that had since became a bit stale.

Before this PR gets merged in we must release libddprof 0.6.0 on rubygems.org: https://github.com/DataDog/libddprof/pull/59 since older versions will break some of our integration tests (see da925a0b35064b026443e58fe26c414158595d94 for details).

ivoanjo commented:
Converted to draft as we're still waiting on 1.1.0 to go out, and then possible a rebase and maybe pull in a fix or two :)

ivoanjo commented:
I've rebased this PR, did a few cleanups, and disabled actually linking to libddprof so we can merge it in ASAP.

marcotc commented on spec/datadog/profiling/integration_spec.rb:
```diff
@@ -15,8 +15,7 @@
   before do
-    skip 'Profiling is not supported on JRuby' if PlatformHelpers.jruby?
-    skip 'Profiling is not supported on TruffleRuby' if PlatformHelpers.truffleruby?
+    skip_if_profiling_not_supported(self)
```
🧹

ivoanjo commented:
Thanks y'all! Let's get this show on the road :)
